ipa-core/src/protocol/ipa_prf/malicious_security/lagrange.rs

```diff
@@ -0,0 +1,291 @@
+/// the denominators change for different evaluation points
+/// the function multiplies the denominators to `denominators`, i.e. input `F::ONE`
+pub fn compute_lagrange_denominator<F>(evaluation_points: &[F], denominators: &mut [F]) -> ()
+where
+    F: Field + Invert,
+{
+    debug_assert_eq!(evaluation_points.len(), denominators.len());
+    // check that differences are non-zero
+    #[cfg(debug_assertions)]
+    {
+        for i in 0..evaluation_points.len() {
+            for j in 0..evaluation_points.len() {
+                debug_assert_eq!(
+                    (
+                        i,
+                        j,
+                        evaluation_points[i] - evaluation_points[j] != F::ZERO || j == i
+                    ),
+                    (i, j, true)
+                )
+            }
+        }
+    }
+    // compute denominators:
+    for i in 0..evaluation_points.len() {
+        if i > 0 {
+            for j in 0..i {
+                denominators[i] *= evaluation_points[i] - evaluation_points[j];
+            }
+        }
+        if i + 1 < evaluation_points.len() {
+            for j in i + 1..evaluation_points.len() {
+                denominators[i] *= evaluation_points[i] - evaluation_points[j];
+            }
+        }
+        denominators[i] = denominators[i].invert();
+    }
+}
+
+/// computes the lagrange base polynomials
+/// for correctness, the denominators need to be consistent with the evaluation_points
+/// overwrites `base`
+pub fn compute_lagrange_base<F>(
+    points: &[F],
+    evaluation_points: &[F],
+    denominators: &[F],
+    base: &mut Vec<Vec<F>>,
+) -> ()
+where
+    F: Field,
+{
+    debug_assert_eq!(evaluation_points.len(), denominators.len());
+    #[cfg(debug_assertions)]
+    {
+        for base in base.iter() {
+            debug_assert_eq!(evaluation_points.len(), base.len());
+        }
+    }
+    debug_assert_eq!(points.len(), base.len());
+
+    // for each point, compute base polynomials
+    points
+        .iter()
+        .zip(base.iter_mut())
+        .for_each(|(point, base)| {
+            // set up the denominators
+            base.copy_from_slice(denominators);
+            // compute the nominator
+            for i in 0..evaluation_points.len() {
+                if i > 0 {
+                    for j in 0..i {
+                        base[i] *= *point - evaluation_points[j];
+                    }
+                }
+                if i + 1 < evaluation_points.len() {
+                    for j in i + 1..evaluation_points.len() {
+                        base[i] *= *point - evaluation_points[j];
+                    }
+                }
+            }
+        });
+}
+
+/// precomputed lagrange evaluation
+/// evaluates polynomial using base polynomials
+/// multiplies `result` with input `result`, i.e. use input `F::ONE`
+pub fn lagrange_evaluation_precomputed<F>(y: &[F], base: &Vec<Vec<F>>, result: &mut [F]) -> ()
```

Comment:
I am more in favor of using strongly typed objects here. We implemented Lagrange interpolation a while ago, see if that can be a fit here

https://github.com/private-attribution/ipa/pull/32